File: modules/agent/workflows/boulder.go
package workflows

import (
	"context"
	"fmt"

	"github.com/woyin/OrangeCoding/modules/agent"
	"github.com/woyin/OrangeCoding/modules/ai"
	"github.com/woyin/OrangeCoding/modules/core"
	"github.com/woyin/OrangeCoding/modules/tools"
)

// BoulderRecovery detects a stuck agent, resets the context, and retries.
type BoulderRecovery struct {
	provider   ai.AiProvider
	registry   *tools.ToolRegistry
	workDir    string
	maxRetries int
}

// NewBoulderRecovery creates a new BoulderRecovery workflow.
func NewBoulderRecovery(provider ai.AiProvider, registry *tools.ToolRegistry, workDir string, maxRetries int) *BoulderRecovery {
	return &BoulderRecovery{
		provider:   provider,
		registry:   registry,
		workDir:    workDir,
		maxRetries: maxRetries,
	}
}

// BoulderResult holds the output of a boulder recovery attempt.
type BoulderResult struct {
	Attempts   int
	Success    bool
	Duration   float64 // seconds
	FinalError string
}

// Run attempts the task, detecting stuck states and retrying with a fresh context.
func (br *BoulderRecovery) BoulderRecovery(ctx context.Context, task string) (*BoulderResult, error) {
	result := &BoulderResult{}

	for attempt := 0; attempt < br.maxRetries; attempt++ {
		result.Attempts = attempt + 1

		// Fresh context for each attempt
		sid := core.NewSessionId()
		agentCtx := agent.NewAgentContext(sid, br.workDir)
		agentCtx.SetSystemPrompt("You are a resilient agent. Complete the task without getting stuck in loops.")
		agentCtx.AddUserMessage(task)

		executor := agent.NewToolExecutor(br.registry)
		toolDefs := buildWorkflowToolDefs(br.registry)

		loop := agent.NewAgentLoop(core.NewAgentId(), br.provider, executor, agentCtx, agent.DefaultLoopConfig(), toolDefs)

		eventCh := make(chan core.AgentEvent, 100)
		go func() {
			for range eventCh {
			}
		}()

		loopResult, err := loop.Run(ctx, ai.ChatOptions{}, eventCh)
		close(eventCh)

		result.Duration += loopResult.Duration.Seconds()

		if err == nil {
			result.Success = true
			return result, nil
		}

		result.FinalError = err.Error()

		// Check context cancellation
		if ctx.Err() != nil {
			return result, fmt.Errorf("boulder recovery canceled on attempt %d: %w", attempt+1, ctx.Err())
		}
	}

	return result, fmt.Errorf("boulder recovery: all %d attempts exhausted, last error: %s", br.maxRetries, result.FinalError)
}

File: packages/agent/src/workflows/boulder.ts
/**
 * BoulderRecovery detects a stuck agent, resets the context, and retries.
 * Ported from modules/agent/workflows/boulder.go.
 */

import type { AiProvider } from "@orangecoding/ai";
import { SessionId, AgentId } from "@orangecoding/core";
import { ToolRegistry } from "@orangecoding/tools";
import { AgentContext } from "../context.js";
import { ToolExecutor } from "../executor.js";
import { buildToolDefinitions } from "../tool-defs.js";
import { AgentLoop, defaultLoopConfig } from "../loop.js";

/** Result of a boulder-recovery run: attempt count, outcome, and last error. */
export interface BoulderResult {
  attempts: number;
  success: boolean;
  durationMs: number;
  finalError: string;
}

/**
 * Stuck-agent recovery strategy. When an agent fails or loops, BoulderRecovery
 * discards the contaminated context and retries the task from scratch (fresh
 * session + system prompt) up to `_maxRetries` times. The fresh-context reset
 * is the key trick: it breaks tool-call loops that persist within a single
 * conversation. Throws after all attempts are exhausted.
 */
export class BoulderRecovery {
  private _provider: AiProvider;
  private _registry: ToolRegistry;
  private _workDir: string;
  private _maxRetries: number;

  constructor(provider: AiProvider, registry: ToolRegistry, workDir: string, maxRetries: number) {
    this._provider = provider;
    this._registry = registry;
    this._workDir = workDir;
    this._maxRetries = maxRetries;
  }

  /**
   * Retry loop: each attempt spins up a brand-new AgentContext so prior
   * tool-call history cannot trap the agent in the same loop. Returns on the
   * first successful completion; throws after `_maxRetries` failures. Abort
   * is honored between attempts.
   */
  async run(signal: AbortSignal | undefined, task: string): Promise<BoulderResult> {
    const result: BoulderResult = {
      attempts: 0,
      success: false,
      durationMs: 0,
      finalError: "",
    };

    for (let attempt = 0; attempt < this._maxRetries; attempt++) {
      result.attempts = attempt + 1;

      // Fresh context for each attempt
      const sid = SessionId.create();
      const agentCtx = new AgentContext(sid, this._workDir);
      agentCtx.setSystemPrompt("You are a resilient agent. Complete the task without getting stuck in loops.");
      agentCtx.addUserMessage(task);

      const executor = new ToolExecutor(this._registry);
      const toolDefs = buildToolDefinitions(this._registry);
      const loop = new AgentLoop(AgentId.create(), this._provider, executor, agentCtx, defaultLoopConfig(), toolDefs);

      const loopResult = await loop.run({}, null);
      result.durationMs += loopResult.durationMs;

      if (loopResult.stopReason === "completed") {
        result.success = true;
        return result;
      }

      result.finalError = `stop reason: ${loopResult.stopReason}`;

      // Check context cancellation
      if (signal?.aborted) {
        return result;
      }
    }

    throw new Error(`boulder recovery: all ${this._maxRetries} attempts exhausted, last error: ${result.finalError}`);
  }
}
